### PR TITLE
[PIRK-21] - Initial Spark Streaming Responder Implementation

### PR DESCRIPTION
**Pull request by ellisonanne**
The initial Spark Streaming Responder implementation, including corresponding distributed tests. Initial input adaptors only include HDFS and Elasticsearch; additional adaptors (Kafka, etc) will be added as subsequent JIRAs. 

Marking this PR as a WIP until scale testing is completed; functional testing via new streaming tests within the DistributedTestSuite is complete.

Comments/suggestions welcome!

### CODE REVIEW

File: src/main/resources/pirk.properties [hunk 1]
```diff
@@ -28,7 +28,8 @@ local.pirk.properties.dir=/root/
 ##
 ##Spark path for SparkLauncher
 ##
-spark.home = /usr
+#spark.home = /usr
+spark.home=/opt/cloudera/parcels/CDH/lib/spark
```
Comment (tellison) [inline src/main/resources/pirk.properties]: I'm guessing this ```pirk.properties``` was included in the commit by mistake.

File: src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    outputDirExp = outputFile + "_exp";
+
+    queryInput = SystemConfiguration.getProperty("pir.queryInput");
+    String stopListFile = SystemConfiguration.getProperty("pir.stopListFile");
+
+    logger.info("outputFile = " + outputFile + " queryInputDir = " + queryInput + " stopListFile = " + stopListFile + " esQuery = " + esQuery
+        + " esResource = " + esResource);
+
+    // Pull the batchSeconds and windowLength parameters
+    batchSeconds = Long.parseLong(SystemConfiguration.getProperty("pir.sparkstreaming.batchSeconds", "30"));
```
Comment (tellison) [inline src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java]: ```Long``` is a bit generous for a duration in seconds.

File: src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+      throw new IllegalArgumentException("batchSeconds = " + batchSeconds + " must divide windowLength = " + windowLength);
+    }
+    useQueueStream = SystemConfiguration.getProperty("pir.sparkstreaming.useQueueStream", "false").equals("false");
+
+    // Set the necessary configurations
+    SparkConf conf = new SparkConf().setAppName("SparkPIR").setMaster("yarn-cluster");
+    conf.set("es.nodes", SystemConfiguration.getProperty("es.nodes", "none"));
+    conf.set("es.port", SystemConfiguration.getProperty("es.port", "none"));
+    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer");
+    conf.set("spark.streaming.stopGracefullyOnShutdown", SystemConfiguration.getProperty("spark.streaming.stopGracefullyOnShutdown", "false"));
```
Comment (tellison) [inline src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java]: Quite a few places where the ```SystemConfiguration.get*Property``` could be used to help with typing variables.

File: src/main/resources/META-INF/bin-license-notice/LICENSE-bin [hunk 4]
```diff
@@ -0,0 +1,262 @@
+
```
Comment (tellison) [inline src/main/resources/META-INF/bin-license-notice/LICENSE-bin]: I think these license changes are part of another commit.

File: src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+/**
+ * Master class for the PIR query spark streaming application
+ * <p>
+ * NOTE:
+ * <p>
+ * - NOT using Elasticsearch in practice - proved to be some speed issues with ES and Spark that appear to be ES-Spark specific - leave code in anticipation
+ * that the ES-Spark issues resolve...
+ * <p>
+ * - Even if rdd.count() calls are embedded in logger.debug statements, they are computed by Spark. Thus, they are commented out in the code below - uncomment
+ * for rdd.count() debug
```
Comment (tellison) [inline src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java]: Surround with a ```logger.isDebugEnabled()``` check.

File: src/main/resources/pirk.properties (hunk 1 above)
Comment (ellisonanne) [inline src/main/resources/pirk.properties]: Yes, this is for the CDH cluster that I am doing some of the testing on - I will remove it 

(It is, however, an example of how to alternately set the spark.home property for different cluster configurations)

File: src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java (hunk 3 above)
Comment (ellisonanne) [inline src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java]: Yes - I will update it with the new typing.

File: src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java (hunk 2 above)
Comment (ellisonanne) [inline src/main/java/org/apache/pirk/responder/wideskies/spark/streaming/ComputeStreamingResponse.java]: 'batchSeconds' is an input to the Durations class' static seconds method which requires a long (http://spark.apache.org/docs/latest/api/java/) 

But, yes, I do think that a long is a bit of overkill ;)

**Comment by ellisonanne** [comment]
Scale testing is in progress - I would like to go ahead and merge this into master before it gets too far behind... I will update as necessary with any changes that result from scale testing.

**Comment by tellison** [comment]
+1, it is good enough to merge into master and refine from there.

**Comment by ellisonanne** [comment]
Merging now - will keep working on it via JIRA issues and PRs

**Pull request closed without merging**